### PR TITLE
Split files

### PR DESCRIPTION
**Pull request by yoshuawuyts**
Split server.rs and client.rs into separate encoder / decoders for each. Additionally refactored the server encoder loop from a match / break state machine into separate functions. Reviewing this patch is probably best done on a per-commit basis.

This is prep work to further drill down into our out of bound reads issues which I'm still running into. Thanks!

### CODE REVIEW

File: src/client/encode.rs [hunk 1]
```diff
@@ -1,73 +1,16 @@
...
-    let res = decode(stream).await?;
-    log::trace!("< {:?}", &res);
-
-    Ok(res)
-}
 
 /// Encode an HTTP request on the client.
 #[doc(hidden)]
-async fn encode(req: Request) -> http_types::Result<Encoder> {
+pub async fn encode(req: Request) -> http_types::Result<Encoder> {
```
Comment (rylev) [inline src/client/encode.rs]: This probably should be called `encoder` since it doesn't actually encode the request.

File: src/client/decode.rs [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+    ensure!(
+        content_length.is_none() || transfer_encoding.is_none(),
+        "Unexpected Content-Length header"
+    );
+
+    // Check for Transfer-Encoding
+    match transfer_encoding {
+        Some(encoding) if !encoding.is_empty() => {
+            if encoding.last().unwrap().as_str() == "chunked" {
```
Comment (rylev) [inline src/client/decode.rs]: Is combining this `if` with the `if` in the `Some` match less readable? Then we only have one area of control flow

File: src/client/decode.rs [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    R: Read + Unpin + Send + Sync + 'static,
+{
+    let mut reader = BufReader::new(reader);
+    let mut buf = Vec::new();
+    let mut headers = [httparse::EMPTY_HEADER; MAX_HEADERS];
+    let mut httparse_res = httparse::Response::new(&mut headers);
+
+    // Keep reading bytes from the stream until we hit the end of the stream.
+    loop {
+        let bytes_read = reader.read_until(b'\n', &mut buf).await?;
```
Comment (rylev) [inline src/client/decode.rs]: nit: put `\n` in a const

File: src/client/decode.rs [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+        // Prevent CWE-400 DDOS with large HTTP Headers.
+        ensure!(
+            buf.len() < MAX_HEAD_LENGTH,
+            "Head byte length should be less than 8kb"
+        );
+
+        // We've hit the end delimiter of the stream.
+        let idx = buf.len() - 1;
+        if idx >= 3 && &buf[idx - 3..=idx] == b"\r\n\r\n" {
```
Comment (rylev) [inline src/client/decode.rs]: nit: put `\r\n\r\n` in a const

File: src/client/decode.rs (hunk 2 above)
Comment (rylev) [reply src/client/decode.rs]: If so, then you should probably use `if let` like in the server

File: src/server/encode.rs [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+        // we keep track how many bytes of the head and body we've read
+        // in this call of `poll_read`
+        self.bytes_read = 0;
+        match self.state {
+            EncoderState::Start => self.encode_start(cx, buf),
+            EncoderState::Head => self.encode_head(cx, buf),
+            EncoderState::Body => self.encode_body(cx, buf),
+            EncoderState::UncomputedChunked => self.encode_uncomputed_chunked(cx, buf),
+            EncoderState::ComputedChunked => self.encode_computed_chunked(cx, buf),
+            EncoderState::Done => Poll::Ready(Ok(self.bytes_read)),
```
Comment (rylev) [inline src/server/encode.rs]: Might be clearer to just hardcode this to `Ok(0)`

File: src/client/decode.rs (hunk 2 above)
Comment (yoshuawuyts) [reply src/client/decode.rs]: Filed https://github.com/http-rs/http-types/issues/89 to track a structural improvement to this block. The equality check here is too involved still, and I think we can improve.

**Pull request merged**